### PR TITLE
feat: add Joblib model serialization and inference support

### LINKED ISSUE #2042: 💡[FEATURE]: Add Model Serialization Using Joblib for Efficient Startup Profit Prediction
**Is your feature request related to a problem? Please describe.**
Currently, the machine learning model is trained every time the notebook is executed. This increases execution time, makes predictions inefficient, and is not suitable for deployment or integration with web applications. Since the trained model is not persisted, users must retrain it before every prediction, which is unnecessary and time-consuming.

**Describe the solution you'd like**
Refactor the project by separating model training and inference using Joblib.

The proposed implementation includes:
- Train the model once using the existing dataset.
- Serialize and save the trained model as `startup_profit_model.joblib`.
- Create a separate prediction script (`predict.py`) that loads the saved model and performs inference without retraining.
- Save any preprocessing objects (such as encoders or scalers) if required for consistent predictions.
- Organize the project into a cleaner structure by separating datasets, models, training scripts, and inference scripts.
- Update the README with clear instructions for training, saving, loading, and using the serialized model.

**Describe alternatives you've considered**
An alternative is to use Python's built-in Pickle module for model serialization. While Pickle works for saving scikit-learn models, Joblib is generally preferred because it is optimized for NumPy arrays and scikit-learn models, making it more efficient for loading and saving machine learning models.

**Additional context**
Suggested project structure:

Startup_Profit_Prediction/
├── dataset/
│   └── 50_Startups.csv
├── models/
│   └── startup_profit_model.joblib
├── train.py
├── predict.py
├── Startup_Profit_Prediction.ipynb
└── README.md

Benefits:
- Faster predictions without retraining the model.
- Cleaner separation between training and inference.
- Makes the project deployment-ready.
- Easier integration with Flask, FastAPI, or Streamlit applications.
- Improves maintainability and follows better machine learning engineering practices.

name: 💡 Feature Request
about: Improve the ML workflow by saving trained models
title: '💡[FEATURE]: Add Model Serialization Using Joblib for Faster Inference'
labels: 'enhancement'
assignees: ''

### PR DESCRIPTION
**Pull request by Prabhu-E-S**
## Summary

This PR enhances the **Startup Profit Prediction** notebook by integrating **Joblib-based model serialization** while preserving the original machine learning workflow.

The project structure remains unchanged, and the implementation follows the review guidelines by keeping all functionality inside the notebook.

Fixes #2042 
---
## Changes Made

- Added detailed markdown explanations throughout the notebook.
- Improved notebook organization with clear section headings.
- Added **Joblib model serialization** for the trained models.
- Demonstrated loading serialized models within the notebook.
- Added sample inference using the loaded models.
- Executed all notebook cells and saved outputs.
- Updated the README with project overview, workflow, and usage instructions.
- Preserved the original project structure and machine learning pipeline.
---
## Benefits

- Eliminates the need to retrain models before reuse.
- Demonstrates model persistence using Joblib.
- Makes the notebook easier to understand with detailed documentation.
- Keeps the implementation beginner-friendly and educational.
- Preserves the original repository structure as requested by the reviewer.
---
## Testing

- Successfully executed the notebook from start to finish.
- Verified model serialization and loading using Joblib.
- Verified predictions using the loaded models.
- Tested successfully in Google Colab.

### CODE REVIEW

**Comment by Prabhu-E-S** [comment]
@Sibam-Paul Hello, Could u please review this PR and merge?

**Review by Sibam-Paul — CHANGES_REQUESTED** [review]
remove all py files as we need notebook
see the pinned issues for guidelines @Prabhu-E-S

**Comment by Sibam-Paul** [comment]
if there was the dataset present in the original version link if using the` gdown `
and make the link in such a way that it never becomes invalid

**Review by Sibam-Paul — CHANGES_REQUESTED** [review]
make a detailed notebook and all cells must be run with output shown

**Comment by Prabhu-E-S** [comment]
@Sibam-Paul I have done requested changes. Please review it and merge

**Comment by Prabhu-E-S** [comment]
@Sibam-Paul 
Regarding the dataset, I did not use gdown because the original repository already includes 50_startups.csv, and the dataset is also publicly available on Kaggle rather than Google Drive. To preserve the original workflow and repository structure, I continued using the existing dataset file.

Kindly review the updated PR. Thank you!

**Comment by Prabhu-E-S** [comment]
@Sibam-Paul 
Please review it again.
I have done the changes.
Thank you

File: Startup_Profit_Prediction/README.md [hunk 1]
```diff
@@ -1,50 +1,62 @@
-## **Startup Profit Prediction**
-**GOAL**
-
-The goal of this project is to analyse and predict profit of a startup  from features as 'R&D Spend', 'Administration', 'Marketing Spend', 'State' etc.
-
-**DATASET**
-
-Dataset can be downloaded from [here](https://www.kaggle.com/sonalisingh1411/startup50).
```
Comment (Sibam-Paul) [inline Startup_Profit_Prediction/README.md:8]: don't delete the readme completely 

@Prabhu-E-S only add what feature u added extra

restore it

File: Startup_Profit_Prediction/dataset/50_Startups.csv
Comment (Sibam-Paul) [inline Startup_Profit_Prediction/dataset/50_Startups.csv:1]: delete this @Prabhu-E-S

File: Startup_Profit_Prediction/Startup_Profit_Prediction.ipynb
Comment (Sibam-Paul) [inline Startup_Profit_Prediction/Startup_Profit_Prediction.ipynb:1]: Image

fetch the csv form the kaggle as link no the hardcoded one please @Prabhu-E-S

**Review by Sibam-Paul — APPROVED** [review]
@Prabhu-E-S can u come to vc btw @Prabhu-E-S 
pin me in the general

**Comment by Sibam-Paul** [comment]
vc of gssoc discord need to tell u

File: Startup_Profit_Prediction/Startup_Profit_Prediction.ipynb
Comment (Prabhu-E-S) [reply Startup_Profit_Prediction/Startup_Profit_Prediction.ipynb:1]: We can't fetch the csv dataset from kaggle directly as it won't be like a drive.
So keeping the dataset in the repo is my suggestion

Comment (Prabhu-E-S) [reply Startup_Profit_Prediction/Startup_Profit_Prediction.ipynb:1]: @Sibam-Paul

**Comment by Prabhu-E-S** [comment]
> vc of gssoc discord need to tell u

I have sent a msg in discord general chat

File: Startup_Profit_Prediction/Startup_Profit_Prediction.ipynb
Comment (Sibam-Paul) [reply Startup_Profit_Prediction/Startup_Profit_Prediction.ipynb:1]: ` opendatasets `
use this library



@Prabhu-E-S

**Review by Sibam-Paul — CHANGES_REQUESTED** [review]
come to vc after making changes

**Comment by Prabhu-E-S** [comment]
@Sibam-Paul Now I have done all the changes requested by you.
I have removed the dataset from the repo.
I have added the opendataset module to open from kaggle.
I have updated the readme.md as well
Hope now all good :))

**Comment by Prabhu-E-S** [comment]
@Sibam-Paul 
Pls review again

**Review by Sibam-Paul — APPROVED** [review]
LGTM
